### PR TITLE
objc-run: remove relocation hacks

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

These were added in #76394, but may not have been the right thing to do.

-----

For the record, I'm not convinced removing the `post_install` method here is a good idea either, as it could break this formula for Intel users with a non-default prefix.

Keg-relocation mistakenly substitutes `/usr/local` in the following path
```
projectDirPath/build/usr/local/bin/clitest
```
to
```
projectDirPath/build@@HOMEBREW_PREFIX@@/bin/clitest
```
which will resolve to the wrong path in a non-default prefix.

I'm personally inclined to just revert #76394 entirely, at least until https://github.com/Homebrew/brew/issues/11302 is fixed.

CC @MikeMcQuaid

### CODE REVIEW

**Comment by Rylan12** [comment]
> which will resolve to the wrong path in a non-default prefix.
> 
> I'm personally inclined to just revert #76394 entirely, at least until [Homebrew/brew#11302](https://github.com/Homebrew/brew/issues/11302) is fixed.

Yeah, I think I agree. We still have ~400 `bottle :unneeded` formulae in homebrew/core so keeping this as one of them for a little bit until https://github.com/Homebrew/brew/issues/11302 is fixed seems like the best decision to allow for relocatablility.

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by MikeMcQuaid** [comment]
> For the record, I'm not convinced removing the `post_install` method here is a good idea either, as it could break this formula for Intel users with a non-default prefix.

Is it `cellar :any` after the removal? If not, this shouldn't be a problem.

**Comment by carlocab** [comment]
> Is it `cellar :any` after the removal? If not, this shouldn't be a problem.

Does `:any_skip_relocation` count? Tried it locally:
```
❯ brew bottle objc-run
==> Determining objc-run bottle rebuild...
==> Bottling objc-run--1.4.big_sur.bottle.1.tar.gz...
./objc-run--1.4.big_sur.bottle.1.tar.gz
  bottle do
    rebuild 1
    sha256 cellar: :any_skip_relocation, big_sur: "e1b0c25dde3e9a9d31e99955ee2946fc4d8f06274f3dd4d81beb69b71321bdc4"
  end
```

**Comment by Bo98** [comment]
Won't this actively break the script for anyone not on `/usr/local`? It wasn't a problem before because we don't do relocation stuff for `bottle :unneeded`.

**Comment by Rylan12** [comment]
I've never really understood the difference between `:any` and `:any_skip_relocation`, but I think neither is what's needed here.

My understanding of the problem is that right now on Intel machines, the bottles are made to contain `projectDirPath/build@@HOMEBREW_PREFIX@@/bin/clitest`. This means that if they're poured and relocated on another Intel machine that's in a non-default prefix, it will replace `@@HOMEBREW_PREFIX@@` with that machine's prefix which will break whatever internal path it should be referring to.

If it's not `cellar :any`, that means we've now broken relocatability for non-default-prefix users. I guess this might be an okay option because they can still build from source which is no different from what was being done before, but it does seem a little overly inconvenient to me. I'd live with it, though.

**Comment by carlocab** [comment]
> Won't this actively break the script for anyone not on `/usr/local`? It wasn't a problem before because we don't do relocation stuff for `bottle :unneeded`.

Well, yes. That's why I had the `post_install` in the first place.

**Comment by MikeMcQuaid** [comment]
> If it's not `cellar :any`, that means we've now broken relocatability for non-default-prefix users. I guess this might be an okay option because they can still build from source which is no different from what was being done before, but it does seem a little overly inconvenient to me. I'd live with it, though.

Can handle this with a `pour_bottle?` block. Alternatively, use `inreplace` or write a random file with the e.g. cellar location in the prefix to force a non-`cellar :any` bottle.

In general any solution that doesn't push this into `post_install` would be ideal because that slows down the operation for everyone.

https://github.com/Homebrew/brew/issues/11302 is of course gonna be the right thing here.

**Review by Bo98 — DISMISSED** [review]
In that case, I cannot agree that something should be merged with is known to only cause problems for end users without any benefit beyond code style.

Efforts should be redirected to tweak the relocation. Once that's done then the hacks can be removed.

Alternatively, if we really want to remove the hack before fixing the problem, a full revert to `bottle :unneeded` is an alternative temporary solution. This might trip up bottle rebuilds though when we attempt to bottle it again.

**Comment by Bo98** [comment]
> In general any solution that doesn't push this into `post_install` would be ideal because that slows down the operation for everyone.

As long as has the same end result, I'm happy with whatever temporary solution.

**Comment by MikeMcQuaid** [comment]
> In that case, I cannot agree that something should be merged with is known to only cause problems for end users without any benefit beyond code style.

I agree, let's not merge this as-is. I've suggested a few alternate solutions. `post_install` has a non-zero slowdown for users that would be good to avoid (better than this being `cellar :any`, IMO).

**Comment by carlocab** [comment]
Can we just remove the `cellar: :any_skip_relocation` part before pushing it to master?

**Comment by MikeMcQuaid** [comment]
> Can we just remove the `cellar: :any_skip_relocation` part before pushing it to master?

@carlocab That's not really going to scale across maintainers/contributors. The equivalent solution is:
https://github.com/Homebrew/homebrew-core/blob/a32f8fd08629a24537a01af1e57556007a5fbc29/Formula/fontconfig.rb#L23-L28

I'll note in the related issue that these are other potential cases where we're being overly permissive.

**Comment by carlocab** [comment]
> @carlocab That's not really going to scale across maintainers/contributors.

I'm aware; I don't mean for it to. I just think it works for this particular formula, given that it hasn't had a version bump since it was introduced to homebrew/core in 2014. Other formulae that need hacks like this can stay `bottle :unneeded` for now.

That said:

> The equivalent solution is:
> 
> https://github.com/Homebrew/homebrew-core/blob/a32f8fd08629a24537a01af1e57556007a5fbc29/Formula/fontconfig.rb#L23-L28

Happy to try this solution instead.

**Comment by Bo98** [comment]
I'm surprised about that `fontconfig` case. It's not even `cellar: :any`?

**Comment by carlocab** [comment]
> I'm surprised about that `fontconfig` case. It's not even `cellar: :any`?

It was at the time the `pour_bottle?` method was added: https://github.com/Homebrew/homebrew-core/commit/eb4439c5ee5064d0e5bbed09feebf6d15d58f48b

**Comment by Bo98** [comment]
It can probably be removed now then.

**Comment by MikeMcQuaid** [comment]
> It was at the time the `pour_bottle?` method was added: [eb4439c](https://github.com/Homebrew/homebrew-core/commit/eb4439c5ee5064d0e5bbed09feebf6d15d58f48b)

Same deal with `a2ps`: was needed, can now be removed. `ocaml` and `pkg-config` are the only cases where it seems to be legitimately needed (and would be neat if we could handle those).

**Comment by carlocab** [comment]
> Same deal with `a2ps`: was needed, can now be removed. `ocaml` and `pkg-config` are the only cases where it seems to be legitimately needed (and would be neat if we could handle those).

#76640

**Review by MikeMcQuaid — APPROVED** [review]

File: Formula/objc-run.rb
Comment (Bo98) [inline Formula/objc-run.rb:18]: Since we are using an `all` bottle, won't it already have `@@HOMEBREW_PREFIX@@`, and thus be replaced `/opt/homebrew` on ARM?

Comment (carlocab) [reply Formula/objc-run.rb:18]: No, it won't. The reason for the original `inreplace`

https://github.com/Homebrew/homebrew-core/blob/7dd76d8a800c73cb0a8e1488b7863b445158fa68/Formula/objc-run.rb#L15

was because there was an `@@HOMEBREW_PREFIX@@` on Intel, but not on ARM, causing the checksums to differ.

Comment (carlocab) [reply Formula/objc-run.rb:18]: Or rather, it won't anymore after rebuilding the bottles.

Comment (Bo98) [reply Formula/objc-run.rb:18]: Oh right, if we're rebuilding bottles then :+1:

**Review by Bo98 — APPROVED** [review]

**Pull request closed without merging**
